src/lib.rs

```diff
@@ -1237,6 +1406,58 @@ pub async fn build(docker: &Docker, config: &Config) -> Result<(), Error> {
         }
     }
 
+    // Generate {appservice}.yml for appservices running on the host.
+    let generate_appservice_yml =
+        |appservice: &AppServiceConfig, host: &str, path: &Path| -> Result<(), Error> {
+            let parent = path
+                .parent()
+                .ok_or_else(|| anyhow!("Incorrect path {:?}, expected an absolute path", path))?;
+            std::fs::create_dir_all(parent)
+                .with_context(|| format!("Failed to create directory {:?}", parent))?;
+            let mut config = appservice.clone();
+            config.url.set_host(Some(host))?;
+            let writer = std::fs::File::create(&path)
+                .with_context(|| format!("Failed to create file {:?}", path))?;
+            serde_yaml::to_writer(writer, &config)
+                .with_context(|| format!("Failed to write file {:?}", path))?;
+            Ok(())
+        };
+
+    let guest_appservices_root = synapse_root.join("appservices");
+    let host_appservices_root = config.generated_user_files_dir().join("appservices");
+
+    // The (guest) Synapse sees them as running on `host.docker.internal`,
+    // while the (host) AppService see themselves as running on `localhost`.
+    for appservice in &config.appservices.host {
+        // Guest file
+        let guest_path = guest_appservices_root
+            .join(appservice.name.as_ref())
+            .with_extension("yml");
+        generate_appservice_yml(appservice, "host.docker.internal", &guest_path)?;
+        // Host file
+        let host_path = host_appservices_root
+            .join(appservice.name.as_ref())
+            .with_extension("yml");
+        generate_appservice_yml(appservice, "localhost", &host_path)?;
+        eprintln!(
+            "Configuration file for appservice {} has been generated at {:?}",
+            appservice.name, host_path
+        );
+    }
+    // Generate {appservice}.yml for appservices running on the guest.
+    // Both Synapse and the AppService see them as running on `localhost`.
+    for appservice in &config.appservices.guest {
+        let sub_path = Path::new("appservices")
+            .join(appservice.name.as_ref())
+            .with_extension("yml");
+        let path = synapse_root.join(&sub_path);
+        generate_appservice_yml(appservice, "localhost", &path)?;
+        eprintln!(
+            "Configuration file for appservice {} has been generated at guest path {:?}",
+            appservice.name, sub_path
+        );
+    }
```

Comment:
Shouldn't the registration file be generated in the `up` phase like the homeserver config is, so that it is easy to make changes?

Reply:
Would that be easier?

Reply:
Yes, rebuilding means you lose all the data in the homserver

Reply:
Yeah, makes sense. I'll try and fix that.

Reply:
Should be implemented in the latest version.